### PR TITLE
Reorganize acceptance tests

### PR DESCRIPTION
**Pull request by arnolanglade**
**Description (for Contributor and Core Developer)**

During the last test fest we created several in memory. We also moved some builder to the "common" namespace. This PR reorganize our in memory repository and acceptance contexts. In this PR, we regroup them like our future bounded context.
* Catalag (product model, product, category, group, type)
* Channel (currency, locale, channel)
* Structure (attribute, attribute group, attribute opetion, family, family variant)

We added some missing spec for in memory repository.

EDIT: I change my mind, the new structure :
```bash
$ tree tests/back/Acceptance
├── Catalog
│   ├── Category
│   │   └── InMemoryCategoryRepository.php
│   ├── Group
│   │   └── InMemoryGroupRepository.php
│   ├── GroupType
│   │   └── InMemoryGroupTypeRepository.php
│   ├── Product
│   │   └── InMemoryProductRepository.php
│   └── ProductModel
│       └── InMemoryProductModelRepository.php
├── Channel
│   ├── ChannelContext.php
│   ├── Currency
│   │   └── InMemoryCurrencyRepository.php
│   ├── InMemoryChannelRepository.php
│   └── Locale
│       ├── InMemoryLocaleRepository.php
│       └── LocaleContext.php
├── Resources
│   └── config
│       ├── behat
│       │   └── services.yml
│       └── pim
│           └── repositories.yml
├── spec
│   ├── Catalog
│   │   ├── Group
│   │   │   ├── InMemoryGroupRepositorySpec.php
│   │   │   └── Type
│   │   ├── GroupType
│   │   │   └── InMemoryGroupTypeRepositorySpec.php
│   │   ├── Product
│   │   │   ├── InMemoryProductRepositorySpec.php
│   │   │   └── Model
│   │   └── ProductModel
│   │       └── InMemoryProductModelRepositorySpec.php
│   ├── Channel
│   │   ├── Currency
│   │   │   └── InMemoryCurrencyRepositorySpec.php
│   │   ├── InMemoryChannelRepositorySpec.php
│   │   └── Locale
│   │       └── InMemoryLocaleRepositorySpec.php
│   ├── Structure
│   │   ├── Attribute
│   │   │   ├── Group
│   │   │   ├── InMemoryAttributeRepositorySpec.php
│   │   │   └── Option
│   │   ├── AttributeGroup
│   │   │   └── InMemoryAttributeGroupRepositorySpec.php
│   │   ├── AttributeOption
│   │   │   └── InMemoryAttributeOptionRepositorySpec.php
│   │   ├── Family
│   │   │   ├── InMemoryFamilyRepositorySpec.php
│   │   │   └── Variant
│   │   └── FamilyVariant
│   │       └── InMemoryFamilyVariantRepositorySpec.php
│   └── User
│       ├── InMemoryGroupRepositorySpec.php
│       ├── InMemoryRoleRepositorySpec.php
│       └── InMemoryUserRepositorySpec.php
├── Structure
│   ├── Attribute
│   │   └── InMemoryAttributeRepository.php
│   ├── AttributeGroup
│   │   └── InMemoryAttributeGroupRepository.php
│   ├── AttributeOption
│   │   └── InMemoryAttributeOptionRepository.php
│   ├── Family
│   │   └── InMemoryFamilyRepository.php
│   └── FamilyVariant
│       └── InMemoryFamilyVariantRepository.php
└── User
    ├── InMemoryGroupRepository.php
    ├── InMemoryRoleRepository.php
    └── InMemoryUserRepository.php
```

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | yes
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: tests/back/Acceptance/Channel/Currency/InMemoryCurrencyRepository.php [hunk 1]
```diff
@@ -45,6 +46,10 @@ public function findOneByIdentifier($code)
      */
     public function save($currency, array $options = [])
     {
+        if(!$currency instanceof CurrencyInterface) {
+            throw new \InvalidArgumentException('Only user objects are supported.');
```
Comment (skeleton) [inline tests/back/Acceptance/Channel/Currency/InMemoryCurrencyRepository.php]: only currency object

**Comment by jjanvier** [comment]
not easy to understand what's going on without pulling the branch
here is the result

```
$ tree tests/back/Acceptance                                                                                                                                  tests/back/Acceptance
├── Catalog
│   ├── Category
│   │   └── InMemoryCategoryRepository.php
│   ├── Group
│   │   ├── InMemoryGroupRepository.php
│   │   └── Type
│   │       └── InMemoryGroupTypeRepository.php
│   └── Product
│       ├── InMemoryProductRepository.php
│       └── Model
│           └── InMemoryProductModelRepository.php
├── Channel
│   ├── ChannelContext.php
│   ├── Currency
│   │   └── InMemoryCurrencyRepository.php
│   ├── InMemoryChannelRepository.php
│   └── Locale
│       ├── InMemoryLocaleRepository.php
│       └── LocaleContext.php
├── Resources
│   └── config
│       ├── behat
│       │   └── services.yml
│       └── pim
│           └── repositories.yml
├── spec
│   ├── Catalog
│   │   ├── Group
│   │   │   ├── InMemoryGroupRepositorySpec.php
│   │   │   └── Type
│   │   │       └── InMemoryGroupTypeRepositorySpec.php
│   │   └── Product
│   │       ├── InMemoryProductRepositorySpec.php
│   │       └── Model
│   │           └── InMemoryProductModelRepositorySpec.php
│   ├── Channel
│   │   ├── Currency
│   │   │   └── InMemoryCurrencyRepositorySpec.php
│   │   ├── InMemoryChannelRepositorySpec.php
│   │   └── Locale
│   │       └── InMemoryLocaleRepositorySpec.php
│   ├── Structure
│   │   ├── Attribute
│   │   │   ├── Group
│   │   │   │   └── InMemoryAttributeGroupRepositorySpec.php
│   │   │   ├── InMemoryAttributeRepositorySpec.php
│   │   │   └── Option
│   │   │       └── InMemoryAttributeOptionRepositorySpec.php
│   │   └── Family
│   │       ├── InMemoryFamilyRepositorySpec.php
│   │       └── Variant
│   │           └── InMemoryFamilyVariantRepositorySpec.php
│   └── User
│       ├── InMemoryGroupRepositorySpec.php
│       ├── InMemoryRoleRepositorySpec.php
│       └── InMemoryUserRepositorySpec.php
├── Structure
│   ├── Attribute
│   │   ├── Group
│   │   │   └── InMemoryAttributeGroupRepository.php
│   │   ├── InMemoryAttributeRepository.php
│   │   └── Option
│   │       └── InMemoryAttributeOptionRepository.php
│   └── Family
│       ├── InMemoryFamilyRepository.php
│       └── Variant
│           └── InMemoryFamilyVariantRepository.php
└── User
    ├── InMemoryGroupRepository.php
    ├── InMemoryRoleRepository.php
    └── InMemoryUserRepository.php

36 directories, 35 files
```

File: tests/back/Acceptance/Channel/InMemoryChannelRepository.php [hunk 2]
```diff
@@ -42,6 +43,10 @@ public function findOneByIdentifier($code)
      */
     public function save($channel, array $options = [])
     {
+        if(!$channel instanceof ChannelInterface) {
+            throw new \InvalidArgumentException('Only user objects are supported.');
```
Comment (skeleton) [inline tests/back/Acceptance/Channel/InMemoryChannelRepository.php]: only channel

File: tests/back/Acceptance/Channel/Locale/InMemoryLocaleRepository.php [hunk 3]
```diff
@@ -42,6 +43,10 @@ public function findOneByIdentifier($code)
      */
     public function save($locale, array $options = [])
     {
+        if(!$locale instanceof LocaleInterface) {
+            throw new \InvalidArgumentException('Only user objects are supported.');
```
Comment (skeleton) [inline tests/back/Acceptance/Channel/Locale/InMemoryLocaleRepository.php]: locale

File: tests/back/Acceptance/spec/Channel/Currency/InMemoryCurrencyRepositorySpec.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $this->shouldImplement(IdentifiableObjectRepositoryInterface::class);
+    }
+
+    function it_is_a_saver()
+    {
+        $this->shouldImplement(SaverInterface::class);
+    }
+
+    function it_saves_a_role()
```
Comment (skeleton) [inline tests/back/Acceptance/spec/Channel/Currency/InMemoryCurrencyRepositorySpec.php]: a_currency

File: tests/back/Acceptance/spec/Structure/Family/Variant/InMemoryFamilyVariantRepositorySpec.php [hunk 5]
```diff
@@ -15,7 +15,7 @@ class InMemoryFamilyVariantRepositorySpec extends ObjectBehavior
 {
     function it_is_initializable()
     {
-        $this->shouldHaveType(InMemoryFamilyVariantRepository::class);
+        $this->shouldHaveType(\Akeneo\Test\Acceptance\Structure\Family\Variant\InMemoryFamilyVariantRepository::class);
```
Comment (skeleton) [inline tests/back/Acceptance/spec/Structure/Family/Variant/InMemoryFamilyVariantRepositorySpec.php]: to import

**Comment by jjanvier** [comment]
Personally, I don't like the `Group/Type`, `Attribute/Group`, `Attribute/Option` and `Family/Variant` stuff.
I understand what you do here, you try to create something meaningful with the namespace. I find it really useful in a lot of situations, but here it breaks semantic imo. We don't have "variants" but "family variants". We don't have "options" but "attribute options". Those names can't be cut, even for a "useful namespace".

**Comment by aRn0D** [comment]
@jjanvier I understand your point of view, let's see what the rest of the team think. If they don't like that I will change that.

**Comment by juliensnz** [comment]
Nice PR! I have a small question though: is it normal that we use catalog here and not Pim/Catalog? It could be cool to follow the future folder structure even if it creates empty folders no? If we merge it like that, I fear that it will create another version of the "ideal folder structure". If you don't agree, I'm very interested of why :)

**Pull request closed without merging**